Fix: Reward Check Issue || Added: Manual Control Option, Removal of Illegal Actions

J-onah opened the pull request:
Fix:
- Fixed Issue of Reward Check for Function check_if_waiting_beside_bomb_range() Giving Reward when Agent in Bomb Range.

Addons:
- Set Ability to Manual Control (Change Global Variable IS_MANUAL_CONTROL at the top of the Trainer class accordingly) 
- Added Removal of Illegal Actions when Predicting Next Action and when Randomly Picking Actions 
- Added Ability to Toggle Between Checking Illegal Actions and not Checking

J-onah committed:
Fix: Reward Check Issue || Added: Manual Control Option, Removal of Illegal Actions, and Diff. Bomb Frames Grid Values

Fix:
- Fixed Issue of Reward Check for Function check_if_waiting_beside_bomb_range() Giving Reward when Agent in Bomb Range.

Addons:
- Set Ability to Manual Control (Change Global Variable IS_MANUAL_CONTROL at the top of the Trainer class accordingly)
- Added Removal of Illegal Actions when Predicting Next Action and when Randomly Picking Actions
- Added Grid Values for Different Bomb Frames, ie, Grid Values for Bomb Just About to Explode and Bomb Immediately Before Explosion

milselarch reviewed (commented):
running training.py fails as well

```
pygame 2.5.2 (SDL 2.28.2, Python 3.10.14)
Hello from the pygame community. https://www.pygame.org/contribute.html
Traceback (most recent call last):
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/training.py", line 15, in <module>
    trainer.train()
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/trainer.py", line 110, in train
    action = self.agent.act(state, illegal_actions=self.env.get_illegal_actions())
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/dqn.py", line 81, in act
    raw_prediction = self.model.forward(state)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/models/DQN_3D.py", line 84, in forward
    x = self.fc_layers(x)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/milselarch/projects/SUTD/50-021AI/bomberman_rl_jonah/venv/lib/python3.10/site-packages/torch/nn/modules/linear.py", line 116, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x4840 and 2904x256)
```

milselarch reviewed (commented):
also should add a setting to turn off the checking of "illegal" actions

J-onah committed:
Fix: Made Changes According to Comments

Fix:
- Used snake_case Instead of camelCase for Action Space Index Dictionary
- Reduced Conditions Check Length
- Removed Grid Values for Last Two Bomb Frames
- Added Setting Toggle Ability to Toggle Between Checking for Illegal Actions and not Checking

J-onah committed:
Fix: Made Changes According to Comments

J-onah committed:
Removed Changing of Bomb Grid Values to Grid Values for Last Two Bomb Frames

J-onah committed:
Merge branch '110424-merge-with-main' of https://github.com/milselarch/bomberman-RL into 110424-merge-with-main

J-onah committed:
Fix: Illegal Actions Check Correction for Left and Right Grid Indexing
